pkg/autoupdate/findnew_test.go

```diff
@@ -253,6 +295,138 @@ func setupTestDir(t *testing.T, stage setupState) (string, string, func()) {
 		require.NoError(t, os.Chmod(filepath.Join(updatesDir, n, binaryName), 0755))
 	}
 
+	if stage <= executableUpdates {
+		return tmpDir, binaryName, cleanupFunc
+	}
+
+	for _, n := range []string{"5", "1"} {
+		updatedBinaryPath := filepath.Join(updatesDir, n, binaryName)
+		require.NoError(t, copyFile(updatedBinaryPath, binaryPath, true), "copy & truncate executable")
+	}
+
 	return tmpDir, binaryName, cleanupFunc
 
 }
+
+// copyFile copies a file from srcPath to dstPath. If truncate is set,
+// only half the file is copied. (This is a trivial wrapper to
+// simplify setting up test cases)
+func copyFile(dstPath, srcPath string, truncate bool) error {
+	src, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	dst, err := os.Create(dstPath)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	if !truncate {
+		if _, err := io.Copy(dst, src); err != nil {
+			return err
+		}
+	} else {
+		stat, err := src.Stat()
+		if err != nil {
+			return errors.Wrap(err, "statting srcFile")
+		}
+
+		if _, err = io.CopyN(dst, src, stat.Size()/2); err != nil {
+			return errors.Wrap(err, "statting srcFile")
+		}
+	}
+
+	return nil
+}
+
+func TestCheckExecutable(t *testing.T) {
+	t.Parallel()
+	var tests = []struct {
+		testName    string
+		expectedErr bool
+	}{
+		{
+			testName:    "exit0",
+			expectedErr: false,
+		},
+		{
+			testName:    "exit1",
+			expectedErr: false,
+		},
+		{
+			testName:    "exit2",
+			expectedErr: false,
+		},
+		{
+			testName:    "sleep",
+			expectedErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			err := checkExecutable(context.TODO(), os.Args[0], "-test.run=TestHelperProcess", "--", tt.testName)
+			if tt.expectedErr {
+				require.Error(t, err, tt.testName)
+			} else {
+				require.NoError(t, err, tt.testName)
+			}
+		})
+
+	}
+}
+
+func TestCheckExecutableTruncated(t *testing.T) {
+	t.Parallel()
+
+	// First make a broken truncated binary. Lots of setup for this.
+	truncatedBinary, err := ioutil.TempFile("", "test-autoupdate-check-executable-truncation")
+	require.NoError(t, err, "make temp file")
+	defer os.Remove(truncatedBinary.Name())
+	truncatedBinary.Close()
+
+	require.NoError(t, copyFile(truncatedBinary.Name(), os.Args[0], true), "copy executable")
+	require.NoError(t, os.Chmod(truncatedBinary.Name(), 0755))
+
+	require.Error(t,
+		checkExecutable(context.TODO(), truncatedBinary.Name(), "-test.run=TestHelperProcess", "--", "exit0"),
+		"truncated binary")
+}
+
+// TestHelperProcess isn't a real test. It's used as a helper process
+// to make a portableish binary. See
+// https://github.com/golang/go/blob/master/src/os/exec/exec_test.go#L724
+// and https://npf.io/2015/06/testing-exec-command/
```

Comment:
Wow, this took a bit of reading to grok (thanks for providing the explanatory blog post), but it's an interesting/clever way to test exec!

Reply:
This pattern appears in a couple other places in the `launcher` base. I go back and forth on whether it's too clever, or the best there is. I always have to stare at it to remind myself how it works